vm/src/function.rs

```diff
@@ -25,14 +25,61 @@ pub trait IntoPyObject {
     fn into_pyobject(self, vm: &VirtualMachine) -> PyObjectRef;
 }
 
-pub trait IntoPyResult {
-    fn into_pyresult(self, vm: &VirtualMachine) -> PyResult;
+pub trait TryIntoPyObject {
+    fn try_into_pyobject(self, vm: &VirtualMachine) -> PyResult;
+}
+
+impl<T> TryIntoPyObject for T
+where
+    T: IntoPyObject,
+{
+    fn try_into_pyobject(self, vm: &VirtualMachine) -> PyResult {
+        Ok(self.into_pyobject(vm))
+    }
 }
 
 pub trait IntoPyException {
     fn into_pyexception(self, vm: &VirtualMachine) -> PyBaseExceptionRef;
 }
 
+impl IntoPyException for PyBaseExceptionRef {
+    #[inline]
+    fn into_pyexception(self, _vm: &VirtualMachine) -> PyBaseExceptionRef {
+        self
+    }
+}
+
+pub trait IntoPyResult {
+    fn into_pyresult(self, vm: &VirtualMachine) -> PyResult;
+}
+
+impl<T: TryIntoPyObject, E: IntoPyException> IntoPyResult for Result<T, E> {
+    #[inline]
+    fn into_pyresult(self, vm: &VirtualMachine) -> PyResult {
+        self.map_pyerr(vm).and_then(|x| x.try_into_pyobject(vm))
+    }
+}
+
+impl<T: IntoPyObject> IntoPyResult for T {
+    #[inline]
+    fn into_pyresult(self, vm: &VirtualMachine) -> PyResult {
+        Ok(self.into_pyobject(vm))
+    }
+}
+
+pub trait PyErrResultExt {
+    type Ok;
+    fn map_pyerr(self, vm: &VirtualMachine) -> PyResult<Self::Ok>;
+}
+
+impl<T, E: IntoPyException> PyErrResultExt for Result<T, E> {
+    type Ok = T;
+    #[inline]
+    fn map_pyerr(self, vm: &VirtualMachine) -> PyResult<T> {
+        self.map_err(|e| e.into_pyexception(vm))
+    }
+}
```

Comment:
👍

Reply:
Glad you like it! I've been thinking about something like this for a while, to get rid of ".map_err(|e| e.into_pyexception(vm))` __everywhere__

Reply:
I think this is different approach but solving a part of problems I tried to handle in  #3326